pyiron_contrib/generic/filedata.py

```diff
@@ -0,0 +1,332 @@
+"""Generic File Object."""
+
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+import posixpath
+import shutil
+
+from abc import ABC, abstractmethod
+import os
+from functools import lru_cache
+from pathlib import Path
+
+import pandas as pd
+
+from pyiron_base import GenericJob, state
+from pyiron_base import (
+    FileDataTemplate as BaseFileDataTemplate,
+    load_file,
+    FileData as FileDataBase,
+)
+from pyiron_base import HasGroups
+
+try:
+    import ipywidgets as widgets
+    from IPython.display import display
+
+    gui_elements_imported = True
+except ImportError:
+    gui_elements_imported = False
+
+__author__ = "Niklas Siemer"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "0.1"
+__maintainer__ = "Niklas Siemer"
+__email__ = "siemer@mpie.de"
+__status__ = "development"
+__date__ = "Feb 02, 2021"
+
+
+class MetaDataTemplate(ABC):
+    @abstractmethod
+    def to_dict(self):
+        """Provide a dict representation of the meta data."""
+
+    def _repr_html_(self):
+        df = pd.DataFrame(
+            {
+                "Metadata": list(self.to_dict().keys()),
+                "  ": list(self.to_dict().values()),
+            }
+        )
+        df.set_index("Metadata", inplace=True)
+        df[df.isnull()] = ""
+        return df._repr_html_()
+
+
+class MetaData(MetaDataTemplate):
+    def __init__(self, meta_data_dict=None):
+        if meta_data_dict is None:
+            meta_data_dict = {}
+        self._metadata = {key: val for key, val in meta_data_dict.items()}
+
+    def to_dict(self):
+        return self._metadata.copy()
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._metadata})"
+
+
+class FileDataTemplate(BaseFileDataTemplate, ABC):
+    def __init__(self):
+        self._metadata = None
+        if gui_elements_imported:
+            self._output = widgets.Output()
+            self._box = widgets.VBox([self._output])
+
+    @staticmethod
+    def _get_filetype_from_filename(filename):
+        filetype = os.path.splitext(filename)[1]
+        if filetype == "" or filetype == ".":
+            filetype = None
+        else:
+            filetype = filetype[1:]
+        return filetype
+
+    def _get_metadata(self):
+        return self._metadata
+
+    def _set_metadata(self, metadata):
+        if not isinstance(metadata, MetaDataTemplate):
+            self._metadata = MetaData(metadata)
+        else:
+            self._metadata = metadata
```

Comment:
Since I am annoyed by having to put the full
```
@property
def foo(self):
   return self._foo

@foo.setter
def foo(self, val):
   self._foo = val
```
code block in inherited classes.

With
```
@property
def foo(self):
   return self._foo

@foo.setter
def foo(self, val):
   self._set_foo(val)

def _set_foo(self, val)
   self._foo = val
```
I can just redefine the non-property controlled method.
Another option would actually be to use the property directly:

```
def _get_foo(self):
    return self._foo

def _set_foo(self, val):
    self._foo = val

foo = property(_get_foo, _set_foo)
```
This way one would need to redefine `_set_foo` and `foo=property` on inheritance.